README.md
python-tidb-replication
========================

[![Build Status](https://travis-ci.org/HashDataInc/pytidbrep.svg?branch=master)](https://travis-ci.org/HashDataInc/pytidbrep)

Pure Python Implementation of Tidb replication protocol. This allow you to receive event like insert, update, delete with their datas and raw SQL queries.

Use cases
===========

* Tidb to NoSQL database replication
* Tidb to search engine replication
* Invalidate cache when something change in database
* Audit
* Real time analytics

Documentation
==============

TODO

Installation
=============

```
pip install pytidbrep
```

Mailing List
==============

You can get support and discuss about new features on:
https://groups.google.com/forum/#!forum/tidb-user



Examples
=========


This example will dump all replication events to the console:

```python
from pytidbrep import BinLogStreamReader

stream = BinLogStreamReader('/path/to/binlog/dir')

for binlogevent in stream:
    print binlogevent

stream.close()
```

For this SQL sessions:

```sql
CREATE DATABASE test;
use test;
CREATE TABLE test4 (id int NOT NULL AUTO_INCREMENT, data VARCHAR(255), data2 VARCHAR(255), PRIMARY KEY(id));
INSERT INTO test4 (data,data2) VALUES ("Hello", "World");
UPDATE test4 SET data = "World", data2="Hello" WHERE id = 1;
DELETE FROM test4 WHERE id = 1;
```

Output will be:

```
391018326269895930: DDL: CREATE DATABASE test;
391018326269895973: DDL: use test; CREATE TABLE test4 (id int NOT NULL AUTO_INCREMENT, data VARCHAR(255), data2 VARCHAR(255), PRIMARY KEY(id));
391018326269896003: DML: INSERT test.test4: data "Hello", id 1, data2 "World", 
391018326269896009: DML: UPDATE test.test4: data "Hello" => "World", id 1 => 1, data2 "World" => "Hello", 
391018326269896018: DML: DELETE test.test4: data "World", id 1, data2 "Hello", 
```


Tests
========
use `tox` to run unit tests


Changelog
==========
[ChangeLog](https://github.com/HashDataInc/pytidbrep/blob/master/ChangeLog)


Contributors
==============

Major contributor:

* [Zhanwei Wang](https://github.com/wangzw)


Licence
=======
Copyright 2012-2017 HashData Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.



ChangeLog
CHANGES
=======

* do not generate version.py

1.0.0rc2
--------

* fix pip install command
* add build status
* add travis
* improve README
* update changes
* add XidEvent; fix unicode issue
* Change readme and change log
* Refactor code
* first implementation
* Initial commit


tests/test_event.py
# -*- coding:utf-8 -*-

import unittest

from pytidbrep.event import RowsEvent
from pytidbrep.event import read_int32_word


class TestEvent(unittest.TestCase):

    def test_read_int32_word(self):
        row = '112233'.decode("hex")
        v = read_int32_word(row, 3)
        assert 1122867 == v

        row = 'EEDDCC'.decode("hex")
        v = read_int32_word(row, 3)
        v = ~v
        assert 1122867 == v

    def test_parse_int(self):
        pass

    def test_parse_uint(self):
        pass

    def test_read_varint(self):
        pass

    def test_read_uvarint(self):
        pass

    def test_float(self):
        pass

    def test_parse_decimal(self):
        row = '0DFB38D204D2'.decode("hex")
        v = read_int32_word(row, 4)
        assert 234567890 == v

        row = '0E04810DFB38D204D2'.decode("hex")
        v, s = RowsEvent.parse_decimal(row, 0, len(row))
        assert '1234567890.1234' == v
        assert len(row) == s

        row = '0E047EF204C72DFB2D'.decode("hex")
        v, s = RowsEvent.parse_decimal(row, 0, len(row))
        assert '-1234567890.1234' == v
        assert len(row) == s


if __name__ == "__main__":
    # import sys;sys.argv = ['', 'Test.testName']
    unittest.main()


pytidbrep/crc32c.py
#!/usr/bin/env python
#
# Copyright 2007 Google Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""Implementation of CRC-32C checksumming as in rfc3720 section B.4.
See
http://en.wikipedia.org/wiki/Cyclic_redundancy_check for details on CRC-32C.
This code is a manual python translation of c code generated by
pycrc 0.7.1 (http://www.tty1.net/pycrc/). Command line used:
'./pycrc.py --model=crc-32c --generate c --algorithm=table-driven'
"""
import array
CRC_TABLE = (0x00000000L, 0xf26b8303L, 0xe13b70f7L, 0x1350f3f4L, 0xc79a971fL,
             0x35f1141cL, 0x26a1e7e8L, 0xd4ca64ebL, 0x8ad958cfL, 0x78b2dbccL,
             0x6be22838L, 0x9989ab3bL, 0x4d43cfd0L, 0xbf284cd3L, 0xac78bf27L,
             0x5e133c24L, 0x105ec76fL, 0xe235446cL, 0xf165b798L, 0x030e349bL,
             0xd7c45070L, 0x25afd373L, 0x36ff2087L, 0xc494a384L, 0x9a879fa0L,
             0x68ec1ca3L, 0x7bbcef57L, 0x89d76c54L, 0x5d1d08bfL, 0xaf768bbcL,
             0xbc267848L, 0x4e4dfb4bL, 0x20bd8edeL, 0xd2d60dddL, 0xc186fe29L,
             0x33ed7d2aL, 0xe72719c1L, 0x154c9ac2L, 0x061c6936L, 0xf477ea35L,
             0xaa64d611L, 0x580f5512L, 0x4b5fa6e6L, 0xb93425e5L, 0x6dfe410eL,
             0x9f95c20dL, 0x8cc531f9L, 0x7eaeb2faL, 0x30e349b1L, 0xc288cab2L,
             0xd1d83946L, 0x23b3ba45L, 0xf779deaeL, 0x05125dadL, 0x1642ae59L,
             0xe4292d5aL, 0xba3a117eL, 0x4851927dL, 0x5b016189L, 0xa96ae28aL,
             0x7da08661L, 0x8fcb0562L, 0x9c9bf696L, 0x6ef07595L, 0x417b1dbcL,
             0xb3109ebfL, 0xa0406d4bL, 0x522bee48L, 0x86e18aa3L, 0x748a09a0L,
             0x67dafa54L, 0x95b17957L, 0xcba24573L, 0x39c9c670L, 0x2a993584L,
             0xd8f2b687L, 0x0c38d26cL, 0xfe53516fL, 0xed03a29bL, 0x1f682198L,
             0x5125dad3L, 0xa34e59d0L, 0xb01eaa24L, 0x42752927L, 0x96bf4dccL,
             0x64d4cecfL, 0x77843d3bL, 0x85efbe38L, 0xdbfc821cL, 0x2997011fL,
             0x3ac7f2ebL, 0xc8ac71e8L, 0x1c661503L, 0xee0d9600L, 0xfd5d65f4L,
             0x0f36e6f7L, 0x61c69362L, 0x93ad1061L, 0x80fde395L, 0x72966096L,
             0xa65c047dL, 0x5437877eL, 0x4767748aL, 0xb50cf789L, 0xeb1fcbadL,
             0x197448aeL, 0x0a24bb5aL, 0xf84f3859L, 0x2c855cb2L, 0xdeeedfb1L,
             0xcdbe2c45L, 0x3fd5af46L, 0x7198540dL, 0x83f3d70eL, 0x90a324faL,
             0x62c8a7f9L, 0xb602c312L, 0x44694011L, 0x5739b3e5L, 0xa55230e6L,
             0xfb410cc2L, 0x092a8fc1L, 0x1a7a7c35L, 0xe811ff36L, 0x3cdb9bddL,
             0xceb018deL, 0xdde0eb2aL, 0x2f8b6829L, 0x82f63b78L, 0x709db87bL,
             0x63cd4b8fL, 0x91a6c88cL, 0x456cac67L, 0xb7072f64L, 0xa457dc90L,
             0x563c5f93L, 0x082f63b7L, 0xfa44e0b4L, 0xe9141340L, 0x1b7f9043L,
             0xcfb5f4a8L, 0x3dde77abL, 0x2e8e845fL, 0xdce5075cL, 0x92a8fc17L,
             0x60c37f14L, 0x73938ce0L, 0x81f80fe3L, 0x55326b08L, 0xa759e80bL,
             0xb4091bffL, 0x466298fcL, 0x1871a4d8L, 0xea1a27dbL, 0xf94ad42fL,
             0x0b21572cL, 0xdfeb33c7L, 0x2d80b0c4L, 0x3ed04330L, 0xccbbc033L,
             0xa24bb5a6L, 0x502036a5L, 0x4370c551L, 0xb11b4652L, 0x65d122b9L,
             0x97baa1baL, 0x84ea524eL, 0x7681d14dL, 0x2892ed69L, 0xdaf96e6aL,
             0xc9a99d9eL, 0x3bc21e9dL, 0xef087a76L, 0x1d63f975L, 0x0e330a81L,
             0xfc588982L, 0xb21572c9L, 0x407ef1caL, 0x532e023eL, 0xa145813dL,
             0x758fe5d6L, 0x87e466d5L, 0x94b49521L, 0x66df1622L, 0x38cc2a06L,
             0xcaa7a905L, 0xd9f75af1L, 0x2b9cd9f2L, 0xff56bd19L, 0x0d3d3e1aL,
             0x1e6dcdeeL, 0xec064eedL, 0xc38d26c4L, 0x31e6a5c7L, 0x22b65633L,
             0xd0ddd530L, 0x0417b1dbL, 0xf67c32d8L, 0xe52cc12cL, 0x1747422fL,
             0x49547e0bL, 0xbb3ffd08L, 0xa86f0efcL, 0x5a048dffL, 0x8ecee914L,
             0x7ca56a17L, 0x6ff599e3L, 0x9d9e1ae0L, 0xd3d3e1abL, 0x21b862a8L,
             0x32e8915cL, 0xc083125fL, 0x144976b4L, 0xe622f5b7L, 0xf5720643L,
             0x07198540L, 0x590ab964L, 0xab613a67L, 0xb831c993L, 0x4a5a4a90L,
             0x9e902e7bL, 0x6cfbad78L, 0x7fab5e8cL, 0x8dc0dd8fL, 0xe330a81aL,
             0x115b2b19L, 0x020bd8edL, 0xf0605beeL, 0x24aa3f05L, 0xd6c1bc06L,
             0xc5914ff2L, 0x37faccf1L, 0x69e9f0d5L, 0x9b8273d6L, 0x88d28022L,
             0x7ab90321L, 0xae7367caL, 0x5c18e4c9L, 0x4f48173dL, 0xbd23943eL,
             0xf36e6f75L, 0x0105ec76L, 0x12551f82L, 0xe03e9c81L, 0x34f4f86aL,
             0xc69f7b69L, 0xd5cf889dL, 0x27a40b9eL, 0x79b737baL, 0x8bdcb4b9L,
             0x988c474dL, 0x6ae7c44eL, 0xbe2da0a5L, 0x4c4623a6L, 0x5f16d052L,
             0xad7d5351L, )
CRC_INIT = 0
_MASK = 0xFFFFFFFFL


def crc_update(crc, data):
    """Update CRC-32C checksum with data.
  Args:
    crc: 32-bit checksum to update as long.
    data: byte array, string or iterable over bytes.
  Returns:
    32-bit updated CRC-32C as long.
  """
    if type(data) != array.array or data.itemsize != 1:
        buf = array.array("B", data)
    else:
        buf = data
    crc = crc ^ _MASK
    for b in buf:
        table_index = (crc ^ b) & 0xff
        crc = (CRC_TABLE[table_index] ^ (crc >> 8)) & _MASK
    return crc ^ _MASK


def crc_finalize(crc):
    """Finalize CRC-32C checksum.
  This function should be called as last step of crc calculation.
  Args:
    crc: 32-bit checksum as long.
  Returns:
    finalized 32-bit checksum as long
  """
    return crc & _MASK


def crc32c(data):
    """Compute CRC-32C checksum of the data.
  Args:
    data: byte array, string or iterable over bytes.
  Returns:
    32-bit CRC-32C checksum of data as long.
  """
    return crc_finalize(crc_update(CRC_INIT, data))


pytidbrep/expcetion.py
# -*- coding: utf-8 -*-

# Copyright (c) 2016 HashData Inc.
# All Rights Reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.


class ReplicationError(Exception):
    pass


class EofError(ReplicationError):
    pass


class UnexpectedEofError(ReplicationError):
    pass


class MagicNotMatchError(ReplicationError):
    pass


class CrcNotMatcheError(ReplicationError):
    pass


class UnknownBinlogType(ReplicationError):
    pass


class UnknownDMLType(ReplicationError):
    pass


class InvalidRowData(ReplicationError):
    pass


pytidbrep/event.py
# -*- coding: utf-8 -*-

# Copyright (c) 2016 HashData Inc.
# All Rights Reserved.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.


from __future__ import unicode_literals

from datetime import timedelta
import struct

from pytidbrep.expcetion import InvalidRowData
from pytidbrep.pb_binlog.binlog_pb2 import DDL
from pytidbrep.pb_binlog.binlog_pb2 import DML
from pytidbrep.pb_binlog.binlog_pb2 import Delete
from pytidbrep.pb_binlog.binlog_pb2 import Insert
from pytidbrep.pb_binlog.binlog_pb2 import Update
from pytidbrep.pb_binlog.binlog_pb2 import column

SIGNMASK = 0x8000000000000000

DIGITS_PER_WORD = 9  # A word holds 9 digits.
WORD_SIZE = 4  # A word is 4 bytes int32.
DIG2BYTES = [0, 1, 1, 2, 2, 3, 3, 4, 4, 4]

TYPE_BIT = 16
TYPE_TINY = 1
TYPE_SHORT = 2
TYPE_INT24 = 9
TYPE_LONG = 3
TYPE_LONGLONG = 8
TYPE_FLOAT = 4
TYPE_DOUBLE = 5
TYPE_DECIMAL = 0
TYPE_NEWDECIMAL = 0xf6

TYPE_DATE = 10
TYPE_NEWDATE = 14
TYPE_DATETIME = 12
TYPE_TIMESTAMP = 7
TYPE_YEAR = 13
TYPE_DURATION = 11

TYPE_VARCHAR = 15
TYPE_VARSTRING = 0xfd
TYPE_STRING = 0xfe
TYPE_TINYBLOB = 0xf9
TYPE_BLOB = 0xfc
TYPE_MEDIUMBLOB = 0xfa
TYPE_LONGBLOB = 0xfb
TYPE_ENUM = 0xf7
TYPE_SET = 0xf8

TYPE_NULL = 6
TYPE_GEOMETRY = 0xff

MYSQL_TYPE_BIT_STR = "bit"
MYSQL_TYPE_TINYINT = "tinyint"
MYSQL_TYPE_SMALLINT = "smallint"
MYSQL_TYPE_MEDIUMINT = "mediumint"
MYSQL_TYPE_INT = "int"
MYSQL_TYPE_BIGINT = "bigint"
MYSQL_TYPE_DECIMAL = "decimal"
MYSQL_TYPE_FLOAT = "float"
MYSQL_TYPE_DOUBLE = "double"

MYSQL_TYPE_DATE = "date"
MYSQL_TYPE_DATETIME = "datetime"
MYSQL_TYPE_TIMESTAMP = "timestamp"
MYSQL_TYPE_TIME = "time"
MYSQL_TYPE_YEAR = "year"

MYSQL_TYPE_CHAR = "char"
MYSQL_TYPE_VARCHAR = "varchar"
MYSQL_TYPE_BINARY = "binary"
MYSQL_TYPE_VARBINARY = "varbinary"
MYSQL_TYPE_TINYBLOB = "tinyblob"
MYSQL_TYPE_TINYTEXT = "tinytext"
MYSQL_TYPE_BLOB = "blob"
MYSQL_TYPE_TEXT = "text"
MYSQL_TYPE_MEDIUMBLOB = "mediumblob"
MYSQL_TYPE_MEDIUMTEXT = "mediumtext"
MYSQL_TYPE_LONGBLOB = "longblob"
MYSQL_TYPE_LONGTEXT = "longtext"
MYSQL_TYPE_ENUM = "enum"

MYSQL_TYPE_GEOMETRY = "geometry"
MYSQL_TYPE_NULL = "null"
MYSQL_TYPE_VARSTRING = "var_string"
MYSQL_TYPE_UNSPECIFIED = "unspecified"


def get_unicode(s):
    if isinstance(s, unicode):
        return s
    else:
        return s.decode('utf8')


def format_column(t, v):
    _, mysql_type = t

    if v is None:
        return "NULL"
    elif mysql_type in (MYSQL_TYPE_TINYTEXT, MYSQL_TYPE_MEDIUMTEXT,
                        MYSQL_TYPE_LONGTEXT, MYSQL_TYPE_TEXT, MYSQL_TYPE_CHAR,
                        MYSQL_TYPE_VARCHAR):
        return '"%s"' % v
    elif mysql_type in (MYSQL_TYPE_BINARY, MYSQL_TYPE_VARBINARY,
                        MYSQL_TYPE_TINYBLOB, MYSQL_TYPE_MEDIUMBLOB,
                        MYSQL_TYPE_LONGBLOB, MYSQL_TYPE_BLOB):
        return '"%s"' % v.encode('hex')
    elif mysql_type in (MYSQL_TYPE_DATE, MYSQL_TYPE_DATETIME,
                        MYSQL_TYPE_TIMESTAMP, MYSQL_TYPE_TIME,
                        MYSQL_TYPE_YEAR):
        return '"%s"' % str(v)
    else:
        return str(v)


def int2byte(i):
    return struct.pack("!B", i)


def read_be_uint64(buf):
    return struct.unpack(">Q", buf)[0]


def read_int8(buf):
    return struct.unpack(">b", buf)[0]


def read_uint8(buf):
    return struct.unpack(">B", buf)[0]


def read_uvarint(buf):
    '''
     read_uvarint decodes a uint64 from buf and returns that value and the
     number of bytes read (> 0). If an error occurred, the value is 0
     and the number of bytes n is <= 0 meaning:

        n == 0: buf too small
        n  < 0: value larger than 64 bits (overflow)
                  and -n is the number of bytes read
    '''

    x = 0
    s = 0

    for i in range(len(buf)):
        b = read_uint8(buf[i])

        if b < 0x80:
            if i > 9 or i == 9 and b > 1:
                return 0, -(i + 1)  # overflow

            return x | (b << s), i + 1

        x |= (b & 0x7f) << s
        s += 7

    return 0, 0


def read_varint(buf):
    '''
     read_varint decodes an int64 from buf and returns that value and the
     number of bytes read (> 0). If an error occurred, the value is 0
     and the number of bytes n is <= 0 with the following meaning:

        n == 0: buf too small
        n  < 0: value larger than 64 bits (overflow)
                  and -n is the number of bytes read
    '''

    ux, n = read_uvarint(buf)  # ok to continue in presence of error

    x = ux >> 1

    if ux & 1 != 0:
        x = ~x

    return x, n


def read_int32_word(buf, size):
    if read_int8(buf[0]) & 0x80 > 0:
        pad = 0xFF
    else:
        pad = 0

    tmp = bytearray(WORD_SIZE)

    for i in range(WORD_SIZE - size):
        tmp[i] = pad

    offset = WORD_SIZE - size
    for i in range(offset, WORD_SIZE):
        tmp[i] = buf[i - offset]

    x = struct.unpack(">i", str(tmp))[0]

    return x


def decimal_bin_size(precision, frac):
    digits_int = precision - frac
    words_int = digits_int / DIGITS_PER_WORD
    words_frac = frac / DIGITS_PER_WORD
    xint = digits_int - words_int * DIGITS_PER_WORD
    xfrac = frac - words_frac * DIGITS_PER_WORD
    return words_int * WORD_SIZE + DIG2BYTES[xint] + \
        words_frac * WORD_SIZE + DIG2BYTES[xfrac]


class BinLogEvent(object):
    def __init__(self, binlog):
        self.tp = binlog.tp
        self.commit_ts = binlog.commit_ts

    @classmethod
    def type_name(cls, tp):
        if tp == DML:
            return "DML"
        elif tp == DDL:
            return "DDL"
        else:
            return "Unknown"

    def __str__(self):
        return unicode(self).encode('utf-8')

    def __unicode__(self):
        return "%s: %s" % (self.commit_ts, self.type_name(self.tp))


class DDLEvent(BinLogEvent):
    def __init__(self, binlog):
        super(DDLEvent, self).__init__(binlog)
        self._statement = get_unicode(binlog.ddl_query)

    @property
    def statement(self):
        '''Return DDL statement

        @return: A unicode string of DDL statement
        '''
        return self._statement

    def __str__(self):
        return unicode(self).encode('utf-8')

    def __unicode__(self):
        return "%s: %s" % (super(DDLEvent, self).__unicode__(), self.statement)


class XidEvent(BinLogEvent):
    """A COMMIT event
    """

    def __init__(self, ts):
        self.commit_ts = ts

    def __str__(self):
        return unicode(self).encode('utf-8')

    def __unicode__(self):
        return "Transaction at: %s" % (self.commit_ts)


class RowsEvent(BinLogEvent):
    NIL_FLAG = 0
    BYTES_FLAG = 1
    COMPACTBYTES_FLAG = 2
    INT_FLAG = 3
    UINT_FLAG = 4
    FLOAT_FLAG = 5
    DECIMAL_FLAG = 6
    DURATION_FLAG = 7
    VARINT_FLAG = 8
    UVARINT_FLAG = 9
    MAX_FLAG = 250

    def __init__(self, binlog, event):
        super(RowsEvent, self).__init__(binlog)
        self._schema = get_unicode(event.schema_name)
        self._table = get_unicode(event.table_name)
        self._dml_tp = event.tp

    @property
    def schema(self):
        '''Return table's schema name of the event

        @return: A unicode string of schema
        '''
        return self._schema

    @property
    def table(self):
        '''Return table's name of the event

        @return: A unicode string of table name
        '''
        return self._table

    @classmethod
    def dml_type_name(cls, tp):
        if tp == Insert:
            return "INSERT"
        elif tp == Update:
            return "UPDATE"
        elif tp == Delete:
            return "DELETE"
        else:
            return "UNKNOWN"

    def __str__(self):
        return unicode(self).encode('utf-8')

    def __unicode__(self):
        parent = super(RowsEvent, self).__unicode__()
        return "%s: %s %s.%s" % (parent, self.dml_type_name(self._dml_tp),
                                 self.schema, self.table)

    @classmethod
    def parse_int(cls, row, pos, size):
        if size - pos < 8:
            raise InvalidRowData('insufficient bytes to decode value')

        u = read_be_uint64(row[pos:pos + 8])
        v = SIGNMASK ^ u
        return v, 8

    @classmethod
    def parse_uint(cls, row, pos, size):
        if size - pos < 8:
            raise InvalidRowData('insufficient bytes to decode value')

        v = read_be_uint64(row[pos:pos + 8])
        return v, 8

    @classmethod
    def parse_varint(cls, row, pos, size):
        v, n = read_varint(row[pos:pos + 10])  # needs at most 10 bytes
        if n > 0:
            return v, n

        if n < 0:
            raise InvalidRowData("value larger than 64 bits")

        raise InvalidRowData("insufficient bytes to decode value")

    @classmethod
    def parse_uvarint(cls, row, pos, size):
        v, n = read_uvarint(row[pos:pos + 10])  # needs at most 10 bytes
        if n > 0:
            return v, n

        if n < 0:
            raise InvalidRowData("value larger than 64 bits")

        raise InvalidRowData("insufficient bytes to decode value")

    @classmethod
    def parse_float(cls, row, pos, size):
        if size - pos < 8:
            raise InvalidRowData('insufficient bytes to decode value')

        tmp = bytearray(row[pos:pos + 8])
        if tmp[0] & 0x80 > 0:
            tmp[0] &= 0x7F
        else:
            u = struct.unpack(">Q", str(tmp))[0]
            u = ~u
            tmp = struct.pack(">q", u)

        v = struct.unpack(">d", str(tmp))[0]
        return v, 8

    @classmethod
    def parse_bytes(cls, row, pos, size):
        ENC_GROUP_SIZE = 8
        ENC_MARKER = 0xFF
        ENC_PAD = 0x0

        old_pos = pos

        retval = ""
        while True:
            if size - pos < ENC_GROUP_SIZE + 1:
                raise InvalidRowData("insufficient bytes to decode value")

            group = row[pos:pos + ENC_GROUP_SIZE]
            marker = row[pos + ENC_GROUP_SIZE]

            pad_count = ENC_MARKER - marker

            if pad_count > ENC_GROUP_SIZE:
                raise InvalidRowData("invalid marker byte")

            real_roup_size = ENC_GROUP_SIZE - pad_count
            retval += group[:real_roup_size]
            pos += ENC_GROUP_SIZE + 1

            if pad_count != 0:
                pad_byte = ENC_PAD

                for v in group[real_roup_size:]:
                    if v != pad_byte:
                        raise InvalidRowData("invalid padding byte")

                break

        return retval, pos - old_pos

    @classmethod
    def parse_compact_bytes(cls, row, pos, size):
        n, s = cls.parse_varint(row, pos, size)
        if size - pos - s < n:
            raise InvalidRowData(
                "insufficient bytes to decode value, expected length: %s", n)

        return row[pos + s:pos + s + n], s + n

    @classmethod
    def parse_decimal(cls, row, pos, size):
        if size - pos < 3:
            raise InvalidRowData('insufficient bytes to decode value')

        precision = read_int8(row[pos])
        frac = read_int8(row[pos + 1])

        bin_size = decimal_bin_size(precision, frac)

        if size - pos < bin_size + 2:
            raise InvalidRowData("insufficient bytes to decode value")

        bin = row[pos + 2:pos + 2 + bin_size]
        bin_pos = 0

        if read_int8(bin[bin_pos]) & 0x80 > 0:
            negitive = False
            retval = ''
        else:
            negitive = True
            retval = '-'

        bin = bytearray(bin)
        bin[0] ^= 0x80
        bin = str(bin)

        # The number of *decimal* digits before the point.
        digits_int = precision - frac
        # The number of 32bit words before the point.
        words_int = digits_int / DIGITS_PER_WORD
        # The number of leading *decimal* digits not in a word
        leading_digits = digits_int - words_int * DIGITS_PER_WORD

        # The number of 32bit words after the point.
        words_frac = frac / DIGITS_PER_WORD
        # The number of trailing *decimal* digits not in a word
        trailing_digits = frac - words_frac * DIGITS_PER_WORD

        if leading_digits > 0:
            i = DIG2BYTES[leading_digits]
            x = read_int32_word(bin[bin_pos:], i)
            x = ~x if negitive else x
            retval += str(x)
            bin_pos += i

        for i in range(0, words_int * WORD_SIZE, WORD_SIZE):
            x = read_int32_word(bin[bin_pos + i:], WORD_SIZE)
            x = ~x if negitive else x
            retval += str(x)

        bin_pos += words_int * WORD_SIZE

        retval += '.'

        for i in range(0, words_frac * WORD_SIZE, WORD_SIZE):
            x = read_int32_word(bin[bin_pos + i:], WORD_SIZE)
            x = ~x if negitive else x
            retval += str(x)

        bin_pos += words_frac * WORD_SIZE

        if trailing_digits > 0:
            i = DIG2BYTES[trailing_digits]
            x = read_int32_word(bin[bin_pos:], i)
            x = ~x if negitive else x
            retval += str(x)

        return retval, bin_size + 2

    @classmethod
    def parse_duration(cls, row, pos, size):
        v, s = cls.parse_int(row, pos, size)
        d = timedelta(microseconds=v / 1000)

        return d, s

    @classmethod
    def parse_one_column(cls, value):
        pos = 0
        size = len(value)

        if size - pos < 1:
            raise InvalidRowData("Invalid encoded key")

        flag = read_int8(value[pos])
        pos += 1

        if cls.INT_FLAG == flag:
            v, _ = cls.parse_int(value, pos, size)
        elif cls.UINT_FLAG == flag:
            v, _ = cls.parse_uint(value, pos, size)
        elif cls.VARINT_FLAG == flag:
            v, _ = cls.parse_varint(value, pos, size)
        elif cls.UVARINT_FLAG == flag:
            v, _ = cls.parse_uvarint(value, pos, size)
        elif cls.FLOAT_FLAG == flag:
            v, _ = cls.parse_float(value, pos, size)
        elif cls.BYTES_FLAG == flag:
            v, _ = cls.parse_bytes(value, pos, size)
        elif cls.COMPACTBYTES_FLAG == flag:
            v, _ = cls.parse_compact_bytes(value, pos, size)
        elif cls.DECIMAL_FLAG == flag:
            v, _ = cls.parse_decimal(value, pos, size)
        elif cls.DURATION_FLAG == flag:
            v, _ = cls.parse_duration(value, pos, size)
        elif cls.NIL_FLAG == flag:
            v = None
        else:
            raise InvalidRowData("Invalid encoded key")

        return v

    @classmethod
    def parse_insert_and_delete_row(cls, row):
        values = {}
        types = {}

        for c in row:
            col = column.FromString(c)
            tp = read_uint8(col.tp)
            mysql_type = col.mysql_type
            name = get_unicode(col.name)

            value = cls.parse_one_column(col.value)

            if value is not None and \
                    mysql_type in (
                        MYSQL_TYPE_TINYTEXT,
                        MYSQL_TYPE_MEDIUMTEXT,
                        MYSQL_TYPE_LONGTEXT,
                        MYSQL_TYPE_TEXT,
                        MYSQL_TYPE_CHAR,
                        MYSQL_TYPE_VARCHAR
                    ):
                value = value.decode('utf-8')

            values[name] = value
            types[name] = (tp, mysql_type)

        return types, values

    @classmethod
    def parse_update_row(cls, row):
        old_values = {}
        new_values = {}
        types = {}

        for c in row:
            col = column.FromString(c)
            tp = read_uint8(col.tp)
            mysql_type = col.mysql_type
            name = unicode(col.name)

            value = cls.parse_one_column(col.value)
            changed_value = cls.parse_one_column(col.changed_value)

            old_values[name] = value
            new_values[name] = changed_value
            types[name] = (tp, mysql_type)

        return types, old_values, new_values


class DeleteRowsEvent(RowsEvent):
    """This event is trigger when a row in the database is removed

    For each row you have a hash with a single key:
    values which contain the data of the removed line.
    """

    def __init__(self, binlog, event):
        super(DeleteRowsEvent, self).__init__(binlog, event)
        self._types, self._values = self.parse_insert_and_delete_row(event.row)

    @property
    def types(self):
        return self._types

    @property
    def values(self):
        return self._values

    def __str__(self):
        return unicode(self).encode('utf-8')

    def __unicode__(self):
        parent = super(DeleteRowsEvent, self).__unicode__()
        values = self.values
        types = self.types

        s = ''
        for col in values:
            s += "%s %s, " % (col, format_column(types[col], values[col]))

        return "%s: %s" % (parent, s)


class WriteRowsEvent(RowsEvent):
    """This event is triggered when a row in database is added

    For each row you have a hash with a single key:
    values which contain the data of the new line.
    """

    def __init__(self, binlog, event):
        super(WriteRowsEvent, self).__init__(binlog, event)
        self._types, self._values = self.parse_insert_and_delete_row(event.row)

    @property
    def types(self):
        return self._types

    @property
    def values(self):
        return self._values

    def __str__(self):
        return unicode(self).encode('utf-8')

    def __unicode__(self):
        parent = super(WriteRowsEvent, self).__unicode__()
        values = self.values
        types = self.types

        s = ''
        for col in values:
            s += "%s %s, " % (col, format_column(types[col], values[col]))

        return "%s: %s" % (parent, s)


class UpdateRowsEvent(RowsEvent):
    """This event is triggered when a row in the database is changed

    For each row you got a hash with two keys:
        * before_values
        * after_values
    """

    def __init__(self, binlog, event):
        super(UpdateRowsEvent, self).__init__(binlog, event)
        self._types, self._before_values, self._after_values = \
            self.parse_update_row(event.row)

    @property
    def types(self):
        return self._types

    @property
    def before_values(self):
        return self._before_values

    @property
    def after_values(self):
        return self._after_values

    def __str__(self):
        return unicode(self).encode('utf-8')

    def __unicode__(self):
        parent = super(UpdateRowsEvent, self).__unicode__()
        before_values = self.before_values
        after_values = self.after_values
        types = self.types

        s = ''
        for col in before_values:
            s += "%s %s => %s, " % (
                col, format_column(types[col], before_values[col]),
                format_column(types[col], after_values[col]))

        return "%s: %s" % (parent, s)
